File: python/src/train/models/onerank/encoding.py
"""OneRank Transformer block with structured attention masking.

Paper §2.1–§2.2: Structured tokenization with task-specific tokens
and mutual invisibility masking.
"""

from __future__ import annotations

import torch
import torch.nn as nn


def build_attention_mask(num_fields: int, num_tasks: int, device: torch.device) -> torch.Tensor:
    """Build structured attention mask for OneRank.

    Shape [N, N] where N = F + K:
      - Features ↔ Features: 1 (bidirectional)
      - Features → Task tokens: 0
      - Task tokens → Features: 1
      - Task token k → Task token k: 1 (self)
      - Task token k → Task token j (j≠k): 0 (mutual invisibility)

    Returns a float mask: 1.0 = allowed, 0.0 = masked.
    """
    N = num_fields + num_tasks
    mask = torch.zeros(N, N, device=device)
    mask[:num_fields, :num_fields] = 1.0
    mask[num_fields:, :num_fields] = 1.0
    for k in range(num_tasks):
        mask[num_fields + k, num_fields + k] = 1.0
    return mask


class OneRankBlock(nn.Module):
    """One Transformer block for OneRank: Pre-LN MHSA + Pre-LN FFN.

    Uses separate Q/K/V Linear layers (not nn.MultiheadAttention) for
    clean weight naming parity with the Rust Candle implementation:
      attn.q_proj.weight, attn.k_proj.weight, attn.v_proj.weight

    Args:
        d: Model dimension.
        n_heads: Number of attention heads.
        d_ff: FFN intermediate dimension.
        dropout: Dropout rate.
    """

    def __init__(self, d: int, n_heads: int = 8, d_ff: int | None = None, dropout: float = 0.0):
        super().__init__()
        assert d % n_heads == 0, "d must be divisible by n_heads"
        self.d = d
        self.n_heads = n_heads
        self.d_head = d // n_heads
        ffn_dim = d_ff if d_ff is not None else d * 4

        self.norm1 = nn.LayerNorm(d)
        self.q_proj = nn.Linear(d, d)
        self.k_proj = nn.Linear(d, d)
        self.v_proj = nn.Linear(d, d)
        self.o_proj = nn.Linear(d, d)
        self.norm2 = nn.LayerNorm(d)
        self.ffn = nn.Sequential(
            nn.Linear(d, ffn_dim),
            nn.GELU(),
            nn.Linear(ffn_dim, d),
            nn.Dropout(dropout),
        )
        self.dropout = nn.Dropout(dropout)

    def forward(self, x: torch.Tensor, mask_bias: torch.Tensor) -> torch.Tensor:
        """Forward with attention bias (float mask).

        Multi-head dot-product attention with separate Q/K/V projections,
        matching the Candle Rust implementation.

        Args:
            x: [B, N, d]
            mask_bias: [N, N] where 0.0 = allowed, -inf = masked.

        Returns: [B, N, d]
        """
        B, N, _ = x.shape

        residual = x
        x = self.norm1(x)

        q = self.q_proj(x)
        k = self.k_proj(x)
        v = self.v_proj(x)

        # Reshape to multi-head: [B, n_heads, N, d_head]
        q = q.view(B, N, self.n_heads, self.d_head).transpose(1, 2)
        k = k.view(B, N, self.n_heads, self.d_head).transpose(1, 2)
        v = v.view(B, N, self.n_heads, self.d_head).transpose(1, 2)

        # Scaled dot-product
        scores = torch.matmul(q, k.transpose(-2, -1)) * (self.d_head**-0.5)
        scores = scores + mask_bias[None, None, :, :]

        attn = torch.softmax(scores, dim=-1)
        attn = torch.dropout(attn, self.dropout.p, train=self.training)

        out = torch.matmul(attn, v)
        out = out.transpose(1, 2).contiguous().view(B, N, self.d)
        out = self.o_proj(out)
        x = residual + out

        residual = x
        x = self.norm2(x)
        x = self.ffn(x)
        x = residual + x
        return x

File: src/models/onerank/encoding.rs
//! OneRank Transformer block with structured attention masking.
//!
//! Paper §2.1–§2.2: Task-specific token injection with mutual invisibility.

use candle_core::{Module, Result, Tensor};
use candle_nn::{layer_norm, linear, LayerNorm, Linear, VarBuilder};

/// One Transformer block: Pre-LN MHSA + Pre-LN FFN with residual.
pub struct OneRankBlock {
    norm1: LayerNorm,
    q_proj: Linear,
    k_proj: Linear,
    v_proj: Linear,
    o_proj: Linear,
    norm2: LayerNorm,
    ffn_1: Linear,
    ffn_2: Linear,
    #[allow(dead_code)]
    d: usize,
    #[allow(dead_code)]
    n_heads: usize,
    d_head: usize,
}

impl OneRankBlock {
    /// Create a new OneRank block.
    pub fn new(vb: VarBuilder, d: usize, d_ff: usize, n_heads: usize) -> Result<Self> {
        let norm1 = layer_norm(d, 1e-5, vb.pp("norm1"))?;
        let q_proj = linear(d, d, vb.pp("attn.q_proj"))?;
        let k_proj = linear(d, d, vb.pp("attn.k_proj"))?;
        let v_proj = linear(d, d, vb.pp("attn.v_proj"))?;
        let o_proj = linear(d, d, vb.pp("attn.o_proj"))?;
        let norm2 = layer_norm(d, 1e-5, vb.pp("norm2"))?;
        let ffn_1 = linear(d, d_ff, vb.pp("ffn.0"))?;
        let ffn_2 = linear(d_ff, d, vb.pp("ffn.2"))?;
        Ok(Self {
            norm1,
            q_proj,
            k_proj,
            v_proj,
            o_proj,
            norm2,
            ffn_1,
            ffn_2,
            d,
            n_heads,
            d_head: d / n_heads,
        })
    }

    /// Forward with pre-built attention bias mask.
    ///
    /// Args:
    ///   x: [B, N, d]
    ///   mask: [N, N] where 0.0 = allowed, -inf = masked
    pub fn forward(&self, x: &Tensor, mask: &Tensor) -> Result<Tensor> {
        let (b, n, _d) = x.dims3()?;

        // Pre-LN MHSA
        let residual = x;
        let x = self.norm1.forward(x)?;
        let q = self.q_proj.forward(&x)?;
        let k = self.k_proj.forward(&x)?;
        let v = self.v_proj.forward(&x)?;

        // Reshape to multi-head: [B, n_heads, N, d_head]
        let q = q.reshape((b, n, self.n_heads, self.d_head))?.permute((0, 2, 1, 3))?;
        let k = k.reshape((b, n, self.n_heads, self.d_head))?.permute((0, 2, 1, 3))?;
        let v = v.reshape((b, n, self.n_heads, self.d_head))?.permute((0, 2, 1, 3))?;

        // Scaled dot-product: scores [B, n_heads, N, N]
        let scale = (self.d_head as f64).sqrt();
        let mut scores = q.matmul(&k.permute((0, 1, 3, 2))?)?;
        scores = scores.broadcast_div(&Tensor::from_slice(&[scale], (1,), x.device())?)?;

        // Apply mask: broadcast [N, N] → [1, 1, N, N]
        scores = scores.broadcast_add(&mask.unsqueeze(0)?.unsqueeze(0)?)?;

        let attn = candle_nn::ops::softmax(&scores, 3)?;
        let out = attn.matmul(&v)?;
        let x = out.permute((0, 2, 1, 3))?.reshape((b, n, self.d))?;
        let x = self.o_proj.forward(&x)?;
        let x = (residual + x)?;

        // Pre-LN FFN (GELU)
        let residual = &x;
        let x = self.norm2.forward(&x)?;
        let x = self.ffn_1.forward(&x)?;
        let x = x.gelu()?;
        let x = self.ffn_2.forward(&x)?;
        let x = (residual + x)?;

        Ok(x)
    }
}

/// Build structured attention mask for OneRank.
///
/// Shape [N, N] where N = F + K:
///   - Features ↔ Features: 0.0 (allowed)
///   - Features → Task tokens: -inf (masked)
///   - Task tokens → Features: 0.0 (allowed)
///   - Task token k → Task token k: 0.0 (self)
///   - Task token k → Task token j (j≠k): -inf (mutual invisibility)
pub fn build_attention_mask(
    num_fields: usize,
    num_tasks: usize,
    device: &candle_core::Device,
) -> Result<Tensor> {
    let n = num_fields + num_tasks;
    let mut data = vec![-f32::INFINITY; n * n];
    for i in 0..n {
        for j in 0..n {
            let idx = i * n + j;
            if i < num_fields && j < num_fields {
                // Features ↔ Features
                data[idx] = 0.0;
            } else if i >= num_fields && j < num_fields {
                // Task tokens → Features
                data[idx] = 0.0;
            } else if i >= num_fields && j == i {
                // Task token self
                data[idx] = 0.0;
            }
        }
    }
    Tensor::from_slice(&data, (n, n), device)
}
